src/Lykke.AzureStorage/Tables/Decorators/EncryptedTableStorageDecorator.cs

```diff
@@ -0,0 +1,461 @@
+using Lykke.AzureStorage;
+using Lykke.AzureStorage.Cryptography;
+using Lykke.AzureStorage.Tables.Decorators;
+using Lykke.AzureStorage.Tables.Paging;
+using Microsoft.WindowsAzure.Storage.Table;
+
+namespace AzureStorage.Tables.Decorators
+{
+    /// <summary>
+    /// Wrapper on INoSQLTableStorage, uses symmetric encryption.
+    /// Does not support queries (TableQuery<T>) and batches(TableBatchOperation).
+    /// </summary>
+    public class EncryptedTableStorageDecorator<T> : INoSQLTableStorage<T> where T : ITableEntity, new()
+    {
+        private readonly INoSQLTableStorage<T> _storage;
+        private readonly ICryptographicSerializer _serializer;
+        private readonly bool _forceEncrypt;
+        private readonly List<PropertyInfo> _encryptedProperties;
+
+        /// <summary>
+        /// Ctor.
+        /// </summary>
+        /// <param name="storage"></param>
+        /// <param name="serializer"></param>
+        /// <param name="forceEncrypt">If true, checks if existed entires are encrypted, and encrypt them if not.</param>
+        public EncryptedTableStorageDecorator(
+            [NotNull] INoSQLTableStorage<T> storage,
+            [NotNull] ICryptographicSerializer serializer,
+            bool forceEncrypt = true)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+            _forceEncrypt = forceEncrypt;
+
+            var encryptAttribute = typeof(EncryptAttribute);
+            _encryptedProperties = typeof(T).GetProperties().Where(x => Attribute.IsDefined(x, encryptAttribute)).ToList();
+            if (_encryptedProperties.Count == 0)
+                throw new ArgumentException("No properties marked as encrypted.");
+            if (_encryptedProperties.Any(x => x.PropertyType != typeof(string)))
+                throw new ArgumentException($"Only {typeof(string).FullName} type properties can be marked as encrypted.");
+            if (_encryptedProperties.Any(x => x.SetMethod == null || x.GetMethod == null))
+                throw new ArgumentException("Only properties with both Get and Set method allowed. Please check this: " + string.Join(", ",
+                                                _encryptedProperties.Where(x => x.SetMethod == null || x.GetMethod == null).Select(x => x.Name)));
+        }
+
+        public static INoSQLTableStorage<T> Create(INoSQLTableStorage<T> storage, ICryptographicSerializer algo)
+        {
+            return new EncryptedTableStorageDecorator<T>(storage, algo);
+        }
+
+        private T Decrypt(T entity)
+        {
+            if (entity == null)
+            {
+                return default(T);
+            }
+
+            var hasNonEncryptedProperties = false;
+            foreach (var property in _encryptedProperties)
+            {
+                var encrypted = property.GetValue(entity) as string;
```

Comment:
Will be nice to implement it using compiled expressions instead of reflection

Reply:
That's correct. We discussed this with MikhailGerasimov. Theoretically, we can have 20x boost using compiled expressions. But this is a local optimization which give us nothing compared to the whole process. On my PC test of 20M get-set operations was performed in 3.15 vs 0.175 seconds. May be we should use such optimization, but in the end source code will be hard to read.

Reply:
I agree, that this optimization will be not so significant compared to the overall process, but since this is library code, will be better to optimize it, if possible. 

EncryptedTableStorageDecorator code will not be more complicated, if property value reading/writing will be localized in separate classes. Looks like you can even use implementations, that were made for ```AzureTableEntity```: https://github.com/LykkeCity/AzureStorage/blob/master/src/Lykke.AzureStorage/Tables/Entity/PropertyAccess/Factories/PropertyGettersFactory.cs and https://github.com/LykkeCity/AzureStorage/blob/master/src/Lykke.AzureStorage/Tables/Entity/PropertyAccess/Factories/PropertySettersFactory.cs. I think you could add ```AzureTableEntity``` constraint for ```T``` of ```EncryptedTableStorageDecorator<T>``` to make it work.